### PR TITLE
Back port index progress visibility

### PR DESCRIPTION
**Pull request by kvahed**
### Scope & Purpose

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/1428

Backport of index [visibility progress](https://github.com/arangodb/arangodb/pull/18516) on `_api/index`*

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [x] Backport for 3.10: this PR

#### Related Information

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1428
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by jsteemann — APPROVED** [review]
LGTM

**Review by jsteemann — APPROVED** [review]
LGTM

File: arangod/VocBase/Methods/Indexes.cpp [hunk 1]
```diff
@@ -171,7 +171,7 @@ arangodb::Result Indexes::getAll(
       }
     }
     tmp.close();
-
+    std::cout << tmp.toJson() << std::endl;
```
Comment (jsteemann) [inline arangod/VocBase/Methods/Indexes.cpp:174]: Must be removed before merge!
```suggestion
```

File: tests/js/common/shell/shell-index.js [hunk 2]
```diff
@@ -55,6 +55,22 @@ function IndexSuite() {
...
+      for (var i = 0; i < 10000; ++i) {
+        docs.push({name : "name" + i});
+      }
+      c.insert(docs);
+      internal.debugSetFailAt("fillIndex::pause");
+      c.ensureIndex({ type: "persistent", fields:["name"]});
+      console.error("hu");
+      require('internal').sleep(1);
+      console.error("ha");
+      internal.debugSetFailAt("fillIndex::unpause");
```
Comment (jsteemann) [inline tests/js/common/shell/shell-index.js:71]: This is not a "real" test, in the sense of that does not have any conditions under which it will fail
So it executes something that one can attend and watch, but for automated tests it is not so helpful.
Apart from that, the test suite will be execute against a single server and cluster alike, and in the cluster it will likely not be any useful, because the failure points would need to be propagated to all DB servers for this (which I guess is not what is done here).

**Review by jsteemann — CHANGES_REQUESTED** [review]
See my last 2 comments

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 3]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
       return RestStatus::DONE;
     }
 
+    std::stringstream ap;
+    ap << "Plan/Collections/" << _vocbase.name() << "/" << coll->planId()
+       << "/indexes";
+    auto& ac = _vocbase.server().getFeature<ClusterFeature>().agencyCache();
+    auto [plannedIndexes, idx] = ac.get(ap.str());
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:204]: I think we must not execute this code if we are not on a coordinator. At least it should not be executed on a single server, most likely also not on DB servers.

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 4]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
             VPackValue(static_cast<int>(ResponseCode::OK)));
-    tmp.add("indexes", indexes.slice());
+    tmp.add(VPackValue("indexes"));
+    {
+      VPackArrayBuilder guard(&tmp);
+      for (auto const& i : VPackArrayIterator(indexes.slice())) {
+        tmp.add(i);
+      }
+      for (auto const& pi : VPackArrayIterator(plannedIndexes->slice())) {
+        if (pi.hasKey("isBuilding") && pi.get("isBuilding").getBool()) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:219]: can be simplified to just
```suggestion
        if (pi.hasKey("isBuilding").isTrue()) {
```

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 5]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
+          VPackObjectBuilder o(&tmp);
+          for (auto const& source : VPackObjectIterator(pi)) {
+            tmp.add(source.key.copyString(), source.value);
+          }
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:223]: can be improved to
```suggestion
          for (auto source : VPackObjectIterator(pi, /*useSequentialIteration*/ true)) {
            tmp.add(source.key.stringView(), source.value);
          }
```

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 6]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
+          double progress = 0;
+          double success = 0;
+          auto const shards = coll->shardIds();
+          auto const body = VPackBuffer<uint8_t>();
+          auto* pool =
+              coll->vocbase().server().getFeature<NetworkFeature>().pool();
+          std::vector<Future<network::Response>> futures;
+          futures.reserve(shards->size());
+          std::string const prefix = "/_api/index/";
+          network::RequestOptions reqOpts;
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:234]: The requests will always be executed inside the system database, which will not work if the underlying collection is executed in a different database.
The setup of the RequestOptions is missing the `database` attribute here, i.e. `reqOptions.database = _vocbase.name();`

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 7]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
+          reqOpts.param("withHidden", "true");
+          // best effort. only displaying progress
+          reqOpts.timeout = network::Timeout(10.0);
+          for (auto const& shard : *shards) {
+            std::string const url = prefix + shard.first + "/" + iid;
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:239]: Better alternative:
```suggestion
            std::string const url = absl::StrCat(prefix, shard.first, "/", pi.get("id").stringView());
```

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 8]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
+                reqOpts));
+          }
+          for (Future<network::Response>& f : futures) {
+            network::Response const& r = f.get();
+
+            // Only best effort accounting. If something breaks here, we just
+            // ignore the output. Account for what we can and move on.
+            if (r.fail()) {
+              LOG_TOPIC("afde4", INFO, Logger::CLUSTER)
+                  << "Communication error while collecting figures for "
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:251]:
```suggestion
                  << "Communication error while fetching index data for "
```

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 9]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
+            if (!resSlice.isObject() ||
+                !resSlice.get(StaticStrings::Error).isBoolean()) {
+              LOG_TOPIC("aabe4", INFO, Logger::CLUSTER)
+                  << "Result of collecting figures for collection "
+                  << coll->name() << " from " << r.destination << " is invalid";
+              continue;
+            }
+            if (resSlice.get(StaticStrings::Error).getBoolean()) {
+              LOG_TOPIC("a4bea", INFO, Logger::CLUSTER)
+                  << "Failed to collect figures for collection " << coll->name()
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:266]:
```suggestion
                  << "Failed to collect index data for collection " << coll->name()
```

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 10]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
+                  << " from " << r.destination;
+              continue;
+            }
+            if (resSlice.get("progress").isNumber()) {
+              progress += resSlice.get("progress").getNumber<double>();
+              success++;
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:272]: I see that the `success` variable is increased here only, but apart from that it is not used.
Do we need to divide the total progress value by the number of shards that replied successfully? I guess if we don't do this, the progress value could exceed 100% easily.

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 11]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
...
+            } else {
+              LOG_TOPIC("aeab4", INFO, Logger::CLUSTER)
+                  << "No progress entry on index " << iid << "  from "
+                  << r.destination << ": " << resSlice.toJson();
+            }
+          }
+          tmp.add("progress", VPackValue(progress));
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:279]:
```suggestion
          if (success > 0) {
            tmp.add("progress", VPackValue(progress / success));
          }
```

File: arangod/RocksDBEngine/RocksDBBuilderIndex.cpp [hunk 12]
```diff
@@ -152,14 +155,38 @@ Result fillIndexSingleThreaded(
...
+            docsProcessed.load(std::memory_order_relaxed) * 100.0 / count;
+        ridx.progress(p);
+        if (progress != nullptr) {
+          (*progress)(p);
+        }
+      }
+
+#ifdef ARANGODB_ENABLE_FAILURE_TESTS
+      TRI_IF_FAILURE("fillIndex::pause") {
+        std::cout << "fillIndex::pause" << std::endl;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBBuilderIndex.cpp:179]: The print statements here should really be removed, even though they are only included inside a failure point.

File: tests/js/common/shell/shell-index.js [hunk 13]
```diff
@@ -55,6 +55,27 @@ function IndexSuite() {
...
+      internal.debugSetFailAt("fillIndex::pause");
+      c.ensureIndex({ type: "persistent", fields:["name"], inBackground: true});
+      //arango.POST("_api/index?collecion=c0l", { type: "persistent", fields:["name"], inBackground: true}, {"x-arango-async": "store"})
+      console.error("hu");
+      require('internal').sleep(1000);
+      console.error("ha");
+      internal.debugSetFailAt("fillIndex::unpause");
+      internal.debugRemoveFailAt("fillIndex::pause");
+      internal.debugRemoveFailAt("fillIndex::unpause");
+      require('internal').sleep(1000);
```
Comment (jsteemann) [inline tests/js/common/shell/shell-index.js:75]: While this test may work when running it manually and watching its output, it won't help us in any automated test runs.

File: arangod/RocksDBEngine/RocksDBBuilderIndex.cpp (hunk 12 above)
Comment (neunhoef) [inline arangod/RocksDBEngine/RocksDBBuilderIndex.cpp:179]: This will eventually have to be removed.

File: arangod/RocksDBEngine/RocksDBBuilderIndex.cpp [hunk 14]
```diff
@@ -152,14 +155,38 @@ Result fillIndexSingleThreaded(
...
+        while (true) {
+          std::this_thread::sleep_for(std::chrono::milliseconds(3000));
+          // TRI_IF_FAILURE("fillIndex::unpause") {
```
Comment (neunhoef) [inline arangod/RocksDBEngine/RocksDBBuilderIndex.cpp:182]: This is not ready. We need the two failure points for controlled unpause.

File: tests/js/common/shell/shell-index.js [hunk 15]
```diff
@@ -55,6 +55,27 @@ function IndexSuite() {
...
+    testIndexProgress : function () {
+      var c = internal.db._create("c0l");
+      var docs = [];
+      for (var i = 0; i < 10000; ++i) {
+        docs.push({name : "name" + i});
+      }
+      c.insert(docs);
+      internal.debugSetFailAt("fillIndex::pause");
+      c.ensureIndex({ type: "persistent", fields:["name"], inBackground: true});
+      //arango.POST("_api/index?collecion=c0l", { type: "persistent", fields:["name"], inBackground: true}, {"x-arango-async": "store"})
```
Comment (neunhoef) [inline tests/js/common/shell/shell-index.js:68]: This isn't finished.

**Review by neunhoef — CHANGES_REQUESTED** [review]
First read done, some parts missing, in particular in the testing.

**Review by neunhoef — CHANGES_REQUESTED** [review]
I have tried this out manually. For indexes with `inBackground::true` it now seems to work in the cluster. Without `inBackground:true` progress is not reported, just the `isBuilding` index, at least, from what I seem to see, progress is at first 0 for a long time, then it grows (for 6 shards) towards the end to 599.999 or so and then it vanishes when the index is ready.
Furthermore, in both cases, the hidden index is (on the coordinator) only reported in the `indexes` attribute and not in the `identifiers` object. This should behave consistently.
In single server, the APi does not seem to work at all currently, I get this response:
```{
  "code": 500,
  "error": true,
  "errorMessage": "VPack error: Expecting Array slice (exception location: /home/neunhoef/ArangoDB310/arangod/GeneralServer/RestHandler.cpp:548). Please reportthis error to arangodb.com",
  "errorNum": 4
}``` I believe this is because the new code must only be executed on coordinators.

All in all, this is not yet finished.

File: arangod/Cluster/ActionBase.cpp [hunk 16]
```diff
@@ -171,7 +171,9 @@ void ActionBase::startStats() {
 
 /// @brief show progress on Action, and when that progress occurred
 void ActionBase::incStats() {
-  ++_progress;
+  auto progress = _progress.load();
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ActionBase.cpp:174]: yes sir

File: arangod/RestHandler/RestIndexHandler.cpp (hunk 10 above)
Comment (kvahed) [reply arangod/RestHandler/RestIndexHandler.cpp]: on response is divided by number of servers.

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 17]
```diff
@@ -193,13 +197,89 @@ RestStatus RestIndexHandler::getIndexes() {
       return RestStatus::DONE;
     }
 
+    std::stringstream ap;
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/RestHandler/RestIndexHandler.cpp]: This was at least the plan originally. Only indexes created in background and only, when isHidden is true. Can of course be changed.

File: arangod/Indexes/Index.h [hunk 18]
```diff
@@ -440,6 +440,13 @@ class Index {
 
   static size_t sortWeight(aql::AstNode const* node);
 
+  void progress(double p) noexcept {
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/Indexes/Index.h:443]: OK, not really necessary.

File: arangod/RestHandler/RestIndexHandler.cpp (hunk 6 above)
Comment (neunhoef) [reply arangod/RestHandler/RestIndexHandler.cpp]: Done in the new place.

File: arangod/Cluster/ActionBase.cpp [hunk 19]
```diff
@@ -171,9 +171,12 @@ void ActionBase::startStats() {
 
 /// @brief show progress on Action, and when that progress occurred
 void ActionBase::incStats() {
-  ++_progress;
+  // OSX clang cannot increment double very rare calls
+  std::mutex m;
```
Comment (neunhoef) [inline arangod/Cluster/ActionBase.cpp:175]: Using a local variable of type std::mutex is nonsense. This protects nothing at all. Use a mutex as a member variable and please use a mutex guard for locking.

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 20]
```diff
@@ -199,7 +205,97 @@ RestStatus RestIndexHandler::getIndexes() {
...
+                } else {
+                  // Obviously, the index is already ready there.
+                  progress += 100.0;
+                  LOG_TOPIC("aeab4", DEBUG, Logger::CLUSTER)
+                      << "No progress entry on index " << iid << "  from "
+                      << r.destination << ": " << resSlice.toJson()
+                      << " index already finished.";
+                }
+              }
+              tmp.add("progress", VPackValue(progress / shards->size()));
```
Comment (neunhoef) [inline arangod/RestHandler/RestIndexHandler.cpp:289]: Let's add this only if `progress` is non-zero, since otherwise we will always see 0 in the cluster when `inBackground` is set to `false`.

File: tests/js/client/shell/shell-index.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+          assertTrue(idx.progress >= progress);
+          progress = idx.progress;
+          print(progress.toPrecision(4) + "%");
+          sleep(0.1);
+        } else {
+          break;
+        }
+      }
+
+      internal.debugRemoveFailAt("fillIndex::pause");
```
Comment (neunhoef) [inline tests/js/client/shell/shell-index.js:107]: These two should go into a finally clause, otherwise they could remain behind if an exception is thrown.

File: tests/js/client/shell/shell-index.js [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+      internal.debugSetFailAt("fillIndex::pause");
+      var idx = { name:"progress", type: "persistent", fields:["name"], inBackground: true};
+      arango.GET(`/_api/index?collection=c0l`);
+
+      var job = arango.POST_RAW(`/_api/index?collection=c0l`, idx,
+                                {"x-arango-async": "store"}).headers["x-arango-async-id"];
+      let count = 0;
+      var progress = 0.0;
+      var idxs = [];
+      while (true) {
```
Comment (neunhoef) [inline tests/js/client/shell/shell-index.js:86]: This whole test will not fail if one day no progress is reported at all. in fact, if would not have failed without this feature. The reason is that no assertion ever in here requires that `progress` is actually set!
We agreed to do this differently: We said that we set the pause failure point and then wait in the test until at least for once we see `progress` with a value between 0 and 100 (exclusive). Then we would release the pause with the second failure point and let it finish. In the meantime we can still check and if we see `progress`, it must be at least as large as the previous time (check non-decreasing). We should also check that `isBuilding` is set if and only if `progress` is set.

File: tests/js/common/shell/shell-index.js [hunk 23]
```diff
@@ -32,6 +32,7 @@ const errors = internal.errors;
 const { helper, versionHas } = require("@arangodb/test-helper");
 const platform = require('internal').platform;
 
+
```
Comment (neunhoef) [inline tests/js/common/shell/shell-index.js:35]: This is an unnecessary change.

File: tests/js/client/shell/shell-index.js [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+      const sleep = require('internal').sleep;
+      for (var j = 0; j < 16; ++j) {
+        docs=[];
+        for (var i = 0; i < 10240; ++i) {
+          docs.push({name : "name" + i});
+        }
+        c.insert(docs);
+      }
+      
+      internal.debugSetFailAt("fillIndex::pause");
```
Comment (neunhoef) [inline tests/js/client/shell/shell-index.js:77]: This will work in single server, because `debugSetFailAt` will set the failure point on the point of contact, which is the coordinator. For the cluster, it is more difficult: We have to first find the endpoints of all dbservers and then use the two argument version in `test-helpers` to activate the failure points there. And likewise in the end switch off the failure points again. I bet that on cluster this only works since the test can never fail.

**Review by neunhoef — CHANGES_REQUESTED** [review]
See further comments, the test is still not all right.

**Review by neunhoef — APPROVED** [review]
LGTM

File: arangod/Cluster/ActionBase.cpp [hunk 25]
```diff
@@ -171,9 +171,10 @@ void ActionBase::startStats() {
 
 /// @brief show progress on Action, and when that progress occurred
 void ActionBase::incStats() {
-  ++_progress;
+  // OSX clang cannot increment double very rare calls
+  std::lock_guard<std::mutex> guard(_progressMutex);
+  _progress = _progress + 1.0;
   _actionLastStat = secs_since_epoch();
```
Comment (jsteemann) [inline arangod/Cluster/ActionBase.cpp:177]: nit: `_actionLastStat` is now updated under the mutex, but read without it.
This is not racy, as `_actionLastStat` is an atomic variable. However, technically it is not necessary to update it under the mutex, and it is slightly confusing to readers of the code.

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 26]
```diff
@@ -199,7 +205,102 @@ RestStatus RestIndexHandler::getIndexes() {
     tmp.add(StaticStrings::Error, VPackValue(false));
     tmp.add(StaticStrings::Code,
             VPackValue(static_cast<int>(ResponseCode::OK)));
-    tmp.add("indexes", indexes.slice());
+
+    if (ServerState::instance()->isCoordinator()) {
+      std::string ap =
+          absl::StrCat("Plan/Collections/", _vocbase.name(), "/",
+                       std::to_string(coll->planId().id()), "/indexes");
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:212]: The extra `std::to_string(number)` is not necessary here: absl::StrCat() handles that automatically.
```suggestion
                       coll->planId().id(), "/indexes");
```

File: arangod/RestHandler/RestIndexHandler.cpp [hunk 27]
```diff
@@ -199,7 +205,102 @@ RestStatus RestIndexHandler::getIndexes() {
...
+                      << "No progress entry on index " << iid << "  from "
+                      << r.destination << ": " << resSlice.toJson()
+                      << " index already finished.";
+                }
+              }
+              if (progress != 0) {
+                // Don't show progress 0, this is in particular relevant
+                // when isBackground is false, in which case no progress
+                // is reported by design.
+                tmp.add("progress", VPackValue(progress / shards->size()));
```
Comment (jsteemann) [inline arangod/RestHandler/RestIndexHandler.cpp:293]: nit: if one of the shards fails to answer, dividing by `shards->size()` may underreport the actual progress.
I don't have a better solution here, except counting the number of shards for which we successfully received a response. But as the progress value may be skewed anyway in this case, it does not matter much.

File: tests/js/client/shell/shell-index.js [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+        for (let i = 0; i < 10240; ++i) {
+          docs.push({name : "name" + i});
+        }
+        c.insert(docs);
+      }
+      
+      try {
+        internal.debugSetFailAt("fillIndex::pause");
+        let idx = { name:"progress", type: "persistent", fields:["name"], inBackground: true};
+        arango.GET(`/_api/index?collection=c0l`);
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:63]: I don't think this HTTP GET has any purpose in the current version of the code, and thus probably can be removed.
```suggestion
```

File: tests/js/client/shell/shell-index.js [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+        let progress = 0.0;
+        let idxs = [];
+        // Wait until the index is there (with withHidden):
+        while (true) {
+          idxs = arango.GET(`/_api/index?collection=c0l&withHidden=true`).indexes;
+          if (idxs.length > 1) {
+            break;
+          }
+          sleep(0.1);
+        }
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:77]: Nit: this will be an unbounded loop if the index doesn't show up for whatever reason.
We could make the loop only try for a certain amount of time or a certain number of attempts to retrieve the index info.

File: tests/js/client/shell/shell-index.js [hunk 30]
```diff
@@ -0,0 +1,109 @@
...
+        let job = arango.POST_RAW(`/_api/index?collection=c0l`, idx,
+                                  {"x-arango-async": "store"}).headers["x-arango-async-id"];
+        let count = 0;
+        let progress = 0.0;
+        let idxs = [];
+        // Wait until the index is there (with withHidden):
+        while (true) {
+          idxs = arango.GET(`/_api/index?collection=c0l&withHidden=true`).indexes;
+          if (idxs.length > 1) {
+            break;
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:74]: We should assert here that the index data we have got actually is for the correct index, by peeking into `idxs[1]` and validating its contents.

File: tests/js/client/shell/shell-index.js [hunk 31]
```diff
@@ -0,0 +1,109 @@
...
+      try {
+        internal.debugSetFailAt("fillIndex::pause");
+        let idx = { name:"progress", type: "persistent", fields:["name"], inBackground: true};
+        arango.GET(`/_api/index?collection=c0l`);
+
+        let job = arango.POST_RAW(`/_api/index?collection=c0l`, idx,
+                                  {"x-arango-async": "store"}).headers["x-arango-async-id"];
+        let count = 0;
+        let progress = 0.0;
+        let idxs = [];
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:69]: nit: `idxs` is a variable that is only used inside the following loop. So it should not be declared outside of the loop.

File: tests/js/client/shell/shell-index.js [hunk 32]
```diff
@@ -0,0 +1,109 @@
...
+          idxs = arango.GET(`/_api/index?collection=c0l&withHidden=true`).indexes;
+          if (idxs.length > 1) {
+            break;
+          }
+          sleep(0.1);
+        }
+
+        let unpaused = false;
+        while (true) {
+          idx = arango.GET(`/_api/index?collection=c0l&withHidden=true`).indexes[1];
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:81]: nit: the `idx` variable that was declared ~20 lines above is used here for some different purpose.
I suggest making this a local variable inside this loop, so the variable scope & purpose become more clear.

File: tests/js/client/shell/shell-index.js [hunk 33]
```diff
@@ -0,0 +1,109 @@
...
+          if (idx.hasOwnProperty("progress")) {
+            assertTrue(idx.progress >= progress);
+            assertTrue(idx.isBuilding);
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:84]: nit: in case the assertions don't hold true, let's already provide all relevant information for debugging.
```suggestion
            assertTrue(idx.progress >= progress, {idx, progress});
            assertTrue(idx.isBuilding, idx);
```

File: tests/js/client/shell/shell-index.js [hunk 34]
```diff
@@ -0,0 +1,109 @@
...
+        while (true) {
+          idxs = arango.GET(`/_api/index?collection=c0l&withHidden=true`).indexes;
+          if (idxs.length > 1) {
+            break;
+          }
+          sleep(0.1);
+        }
+
+        let unpaused = false;
+        while (true) {
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:80]: nit: This loop is only left by `break` once certain conditions become true.
If these conditions do not become true for whatever reason, this loop is unbounded. We should limit the amount of tries or the total time that can spent inside the loop.

File: tests/js/client/shell/shell-index.js [hunk 35]
```diff
@@ -0,0 +1,109 @@
...
+          idx = arango.GET(`/_api/index?collection=c0l&withHidden=true`).indexes[1];
+          if (idx.hasOwnProperty("progress")) {
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:82]: Are there any expectations as to how often we should get into this branch? Right now it may be executed 0 times, and the test will still pass.
Do we expect a minimum amount of times this particular branch is executed? Then we should count the number of invocations and add an assertion for that following the loop body.

File: tests/js/client/shell/shell-index.js [hunk 36]
```diff
@@ -0,0 +1,109 @@
...
+///
+/// @author Kaveh Vahedipour
+////////////////////////////////////////////////////////////////////////////////
+
+const jsunity = require("jsunity");
+const internal = require("internal");
+const arango = internal.arango;
+const errors = internal.errors;
+const { helper, versionHas } = require("@arangodb/test-helper");
+const platform = require('internal').platform;
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:33]: not used
```suggestion
```

File: tests/js/client/shell/shell-index.js [hunk 37]
```diff
@@ -0,0 +1,109 @@
...
+      for (let j = 0; j < 16; ++j) {
+        docs=[];
+        for (let i = 0; i < 10240; ++i) {
+          docs.push({name : "name" + i});
+        }
+        c.insert(docs);
+      }
+      
+      try {
+        internal.debugSetFailAt("fillIndex::pause");
```
Comment (jsteemann) [inline tests/js/client/shell/shell-index.js:61]: suggestion: use `setUp` and `tearDown` functions instead. This will also unclutter the actual test function.

File: arangod/Cluster/ActionBase.cpp (hunk 25 above)
Comment (neunhoef) [reply arangod/Cluster/ActionBase.cpp:177]: Will fix.

File: arangod/RestHandler/RestIndexHandler.cpp (hunk 26 above)
Comment (neunhoef) [reply arangod/RestHandler/RestIndexHandler.cpp:212]: Cool, I did not know this. Will fix.

File: arangod/RestHandler/RestIndexHandler.cpp (hunk 27 above)
Comment (neunhoef) [reply arangod/RestHandler/RestIndexHandler.cpp:293]: This is correct, but unlikely to occur. Dividing only by the number of correct results also has issues. Let's leave it like it is for now.

File: tests/js/client/shell/shell-index.js (hunk 35 above)
Comment (neunhoef) [reply tests/js/client/shell/shell-index.js:82]: The expectation is that we see intermediate progress at least once. I have added an assertion after the loop.

File: tests/js/client/shell/shell-index.js [hunk 38]
```diff
@@ -0,0 +1,109 @@
...
+          } else {
+            assertFalse(idx.hasOwnProperty("isBuilding"));
+            break;
+          }
+        }
+      } finally {
+        internal.debugRemoveFailAt("fillIndex::pause");
+        internal.debugRemoveFailAt("fillIndex::unpause");
+        c.drop();
+      }
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply tests/js/client/shell/shell-index.js:102]: Done.

**Review by jsteemann — APPROVED** [review]
LGTM
